Substrate front-end tutorial (#191)
substrate-developer-hub/substrate-developer-hub.github.io: docs/tutorials/basic-dapp/part-2-query-balances.md

      .then(unsub => { unsubscribe = unsub; })
      .catch(console.error);

    return () => unsubscribe && unsubscribe();
  }, [accounts, api.query.balances, api.query.balances.freeBalance]);
```

We start by creating a `balance` state, that we will initialize at `0`. Then, as we did in the previous part, we will make the query in a React hook, with `useEffect`. However, we will want to display if the balance changes, this `api.query.balances.freeBalance` takes as first argument the account's address you want to request, and you can pass a callback as the second argument. This callback receives the current balance for the given account. This is where we will persist it in the state.

This `query` is a subscription to the storage, this is what will allow us to see the balance change without reloading the page! Because it's a subscription, we must unsubscribe whenever our `Balances` React component unmounts. The `unsubscribe` function is what is actually returned by the `query`. This one must be called in the `return` statement of the `useEffect` that will be called whenever the component unmounts, in our case, it's when we close the tab or refresh out app's page manually.

> Let's strip down the table we return to only show Alice and the account's associated balance.

Author: I think this one is right, although not nice. I'll change the formulation.

 ```js
  // the end of the Balance.js file

  return (
    <>
      <h1>Balances</h1>
      <Table celled striped>
        <Table.Body>
          <Table.Row key={0}>
            <Table.Cell textAlign='right'>{accounts[0].meta.name}</Table.Cell>
            <Table.Cell>{accounts[0].address}</Table.Cell>
            <Table.Cell>{balance}</Table.Cell>
          </Table.Row>
        </Table.Body>
      </Table>
    </>
  );
}
 ```

You can get the working version of this code by visiting `part-2-2` directory:

```bash
cd part-2-2;
yarn;
yarn start;
```

If you run this example, you will get a table with "Alice" account and balance.
![Alice balance](/docs/tutorials/basic-dapp/assets/part-2-2.jpg)

## 2.3 Query and display all balances

Now that we know how querying the state of a module works, we can map through all of our accounts and query the state for each one. The `api` has a handy function to batch multiple queries at the same time, it's called `multi`. You can pass an array of arguments and it will return an array of the result.

> Instead of having just the balance of Alice in the state, we will change it an object mapping `address` with `freeBalance` so that we can use it to display each account's balance. Because the `multi` remains one subscription, the logic for unsubscribing remains the same.

Author: I think that one is right (the "acocunt's") too.

Let's start small and query the balance for one account, Alice.

```js
export default function Balances (props) {
  const { api, keyring } = props;
  const accounts = keyring.getPairs();
  // this state will keep track of Alice's freeBalance
  const [balance, setBalance] = useState('0');

  useEffect(() => {
    let unsubscribe;

    api.query.balances.freeBalance(accounts[0].address, (currentBalance) => {
        setBalance(currentBalance.toString());
      })
      .then(unsub => { unsubscribe = unsub; })
      .catch(console.error);

    return () => unsubscribe && unsubscribe();
  }, [accounts, api.query.balances, api.query.balances.freeBalance]);
```

We start by creating a state variable called `balance`, that we will initialize at `0`. Then, as we did in the previous part, we will make the query in a React hook with `useEffect`. 
We want to display the balance and subscribe to all changes populated by the Substrate node.

**Subscribe**

> The API method `api.query.balances.freeBalance()` takes  an accounts address as its first argument and you can pass an optional callback as the second argument to subscribe to any updates. The callback function is then listening to changes of the `freeBalance` value  of the given account.

Author: ```suggestion
The API method `api.query.balances.freeBalance()` takes  an account address as its first argument and you can pass an optional callback as the second argument to subscribe to any update. The callback function is then listening to changes of the `freeBalance` value  of the given account.
```

We're updating the components state with the values we're getting from the API and that's what allows us to see the balance change without reloading the page! 

**Unsubscribe**
But because it's a subscription, we must unsubscribe whenever our `Balances` React component unmounts. The `unsubscribe` function is what is actually returned by the `query`. This one must be called in the `return` statement of the `useEffect` that will be called whenever the component unmounts. In our case, it's when we close the tab or refresh our applications manually.

> Let's strip down the table we return to only show Alice and the account's associated balance.

Author: ```suggestion
Let's strip down the table we return to only show the balance for Alice.
```
